nanoc-checking/lib/nanoc/checking/runner.rb

```diff
@@ -93,25 +98,74 @@ def run_checks(classes)
         # TODO: remove me
         Nanoc::Core::Compiler.new_for(@site).run_until_reps_built
 
-        checks = []
-        issues = Set.new
-        length = classes.map { |c| c.identifier.to_s.length }.max + 18
-        classes.each do |klass|
-          print format("  %-#{length}s", "Running check #{klass.identifier}… ")
+        checks = classes.map { _1.create(@site) }
+        length = classes.map { _1.identifier.to_s.length }.max
+
+        # Create space in terminal to print status of all checks
+        classes.count.times { puts }
+        $stdout << "\e[#{classes.count}A" # move to top
+
+        # Print all checks (all “pending” for now)
+        puts 'Running checks…'
+        checks.each_with_index do |check, index|
+          log_check(index:, topic: format("  %-#{length}s", check.class.identifier.to_s), state: 'pending')
+        end
+
+        # Run checks in parallel
+        indexes = (0..checks.count - 1).to_a
+        Parallel.each(indexes, in_threads: NUM_THREADS) do |index|
+          check = checks[index]
+
+          log_check(index:, topic: format("  %-#{length}s", check.class.identifier.to_s), state: 'running'.blue)
 
-          check = klass.create(@site)
           check.run
 
-          checks << check
+          state = check.issues.empty? ? 'ok'.green : 'error'.red
+          log_check(index:, topic: format("  %-#{length}s", check.class.identifier.to_s), state:)
+        end
+
+        # Collect issues
+        issues = Set.new
+        checks.each do |check|
           issues.merge(check.issues)
+        end
 
-          # TODO: report progress
+        # Move cursor to below list
+        $stdout << "\e[#{checks.count}B"
```

Comment:
This needs one extra \n to prevent the output from the last test from being overwritten: 
`$stdout << "\e[#{checks.count}B\n"`

Reply:
Good catch! It’s in fact the `puts 'Running checks…'` that’s throwing sand in the works. Moving that earlier fixes the issue.